### DISCUSSION TITLE
Support rmate protocol

### DISCUSSION THREAD

**Discussion by dwt** [zed-industries/zed#23950] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

This was already [mentioned as part of the remote development ticket](https://github.com/zed-industries/zed/issues/5347#issuecomment-1908653412) but the feature is much too important to be sidelined by that discussion.

The `rmate` protocol is very simple, it means listening on localhost on a specific port, and supporting a very small protocol to receive textiles to edit over it. This allows very simple remote work via ssh while editing the files in the comfort of the local editor with local plugins / shortcuts / everything.

* [The original implementation is here](https://github.com/textmate/rmate)
* [A shell only implementation is here](https://github.com/aurora/rmate)
* [The VSCode rmate plugin lists several more versions of rmate](https://marketplace.visualstudio.com/items?itemName=rafaelmaiolla.remote-vscode)

### If applicable, add mockups / screenshots to help present your vision of the feature

<img width="978" alt="image" src="https://github.com/zed-industries/zed/assets/57199/260fdcca-988b-47b8-95a9-ea1c37c10b24">

**Comment by byronrode** [comment]
+1 for this for me.

Textmate used to have a feature to enable it, but as above, VSCode has an extension that allows you to start/stop the server (which just turns off the tunnelling listener I believe). I'd be willing to give this a bash as an extension, perhaps.

**Comment by hywhuangyuwei** [comment]
Wish this too!
Having this will make it incredibly fast to edit remote files without needing to launch the entire app.

**Comment by dwt** [comment]
I would also like to add that it is much easier to implement than full ssh remote editing or any other protocol where part of the editor actually runs on the remote as all that is needed is a port forward to the target machine and some incredibly small programs (that already exist) to be running on the remote.

**Comment by dwt** [comment]
To be honest, I would really like for the rmate-protocoll to get a slight bit more functional, in that it allows opening directories on the remote host, but that is probably best left as an enhancement for later.

**Comment by zuckschwerdt** [comment]
To get things going I made a simple proof-of-concept rmate server for Zed.
It handles incoming rmate TCP connections and then uses Zed with tmp files.

It could be a lot more robust, feature full, safer, prettier. Error handling is a mess. There are unwraps. It will basically panic on any problem.

But it should work just fine as the intended short-term stop-gap until proper support is added in Zed.

Find it at https://github.com/zuckschwerdt/zed-rmate-server with some binaries at https://github.com/zuckschwerdt/zed-rmate-server/releases

**Comment by byronrode** [comment]
This is a great first step. I haven't even had a chance to look at an extension, so thanks!

Sent via Superhuman ( ***@***.*** )

**Comment by byronrode** [comment]
Lovely @zuckschwerdt - just tested this, and it works really nicely

**Comment by zuckschwerdt** [comment]
I've now refactored the protocol handling to be standalone and reusable. Could be a starting point, but it's only tested against the https://github.com/aurora/rmate client so far.

**Comment by byronrode** [comment]
> I've now refactored the protocol handling to be standalone and reusable. Could be a starting point, but it's only tested against the https://github.com/aurora/rmate client so far.

Nice. Will pull the updated version. FYI, in my multiple uses with this yesterday (we lost an entire k8s cluster 😢) I have had no crashes or issues.

The aurora `rmate` client is the only one I ever use, and was based off the `textmate` one IIRC.

**Reply by dwt** [reply]
aurora rmate is the actual client text mate used and uses.

**Comment by zuckschwerdt** [comment]
I can think of one possible issue causing grief: if you save and then close immediately the file change might not be detected before the connection is closed (on detecting Zed closed).
We should re-check for dirty files before closing, but I don't see that in notify-rs. Maybe something like a second of grace period is needed?

**Comment by byronrode** [comment]
In my previous experience I lost some changes due to closing the window soon, so I generally keep it open a while anyway.

In the VS Code plugin I use, it shows a status of sorts in the status bar, something like:

`Starting server…` and then with files it has a saving/saved status for the file/buffer being edited. Generally I keep the window open for a while, and it's usually for configs etc.

**Comment by zuckschwerdt** [comment]
The `mtime` is now tracked and a final check on close should catch all unnoticed changes.

**Comment by maxbrunsfeld** [comment]
I'm converting this to a discussion, since it's not exactly clear what would need to be built into Zed itself, in order to solve the problem.

**Comment by dwt** [comment]
I would love for the Extension built by @zuckschwerdt to be included in the plugin list. 

Then further integration could go one of several ways:
- Built it into the core of Zed, with nice UI to explain how to use and configure `rmate` in SSH and get it onto remote systems. That way it's super easy to activate and will be a superb way to integrate Zed into remote workflows.
- Add https://github.com/zuckschwerdt/zed-rmate-server as a plugin in the plugin browser that autostarts that server

**Reply by byronrode** [reply]
This is great.

Textmate added the rmate server, but way more simple I use `zed-rmate-server` all the time. So I’m all for this.

**Reply by zuckschwerdt** [reply]
Happy to add further code and documentation, refactor things as needed. Just need a direction where this component should go.

**Reply by dwt** [reply]
@zuckschwerdt I have sadly zero Rust or Zed plugin know-how - but that would be my next steps. Do you happen to know how to a achieve this?

**Reply by zuckschwerdt** [reply]
No, I'm not even sure if plugins are allowed to run background activities (the server socket).

**Reply by dwt** [reply]
Hm, Plugins can run language servers, so there has to be something. But I have no clue what and how to achieve any of that.

**Comment by andremacola** [comment]
Hey, thanks for the extension @zuckschwerdt , I used your binary for an app if that's okay.

Although I don't work with Rust, I ended up creating it using Tauri with the help of AI to solve my problem with the rmate protocol and zed with an interface.

For those who are curious:

https://github.com/andremacola/rmate-server-app/releases

Just a reminder that Rust is not a language I master and the app initially resolves my needs.

![Captura de Tela 2025-02-24 às 17 21 14](https://github.com/user-attachments/assets/5ec97d88-6399-4160-848f-64f09a8c26dd)

I'm thinking of adding:

- Option to choose the editor (zed/sublime/vscode...)
- Logs/Transcription of activities

**Reply by zuckschwerdt** [reply]
Sure, great to see that it's useful to you. Interesting approach. Basically a Tauri app without the usual web part to easily get a tray menu? I vaguely remember some generic menu wrapper, maybe I'm thinking of [xbar](https://github.com/matryer/xbar) / [Argos](https://github.com/p-e-w/argos) / [Kargos](https://github.com/lipido/kargos), not sure.

Since Tauri is Rust and my server also you might want to directly include it as dep and run the async server process from Tauri. (…and then perhaps display currently open files in the menu?)

Though ideally the rmate server should be hidden away in a Zed plugin or such.

**Reply by andremacola** [reply]
I agree that it should be native to Zed or an extension, unfortunately it's not possible at the moment. This would also solve the problem of Zed not having a way to open files via terminal when working with remote servers.

**Reply by bbb651** [reply]
You should be able to use the [`tray-icon`](https://crates.io/crates/tray-icon) directly, it's made and used by tauri. It still requires an event loop to run, see examples, so main advantage is reducing the amount of dependencies thus binary size significantly.

**Reply by zuckschwerdt** [reply]
Indeed, using tray-icon and winit a simple tray UI is possible. Is there significant interest in running this as UI instead of a CLI? Give a few 👍 here and I'll add that.

**Reply by andremacola** [reply]
> Indeed, using tray-icon and winit a simple tray UI is possible. Is there significant interest in running this as UI instead of a CLI? Give a few 👍 here and I'll add that.

Sometimes I accidentally close the terminal for some reason, and I avoid leaving processes running in tmux because I've forgotten an active process for days. The app in the tray prevents the process from being terminated and gives me a visual indication that it's active.

I can quickly switch between the editor I want to open as well with an option.

In the end, it's just a gimmick. Ideally, there should be an extension within Zed. If I'm not using the editor, I don't need the protocol to be running.
